src/Raven.Client/Documents/Session/ClusterTransactionSession.cs

```diff
@@ -0,0 +1,165 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Raven.Client.Documents.Operations.CompareExchange;
+using Raven.Client.Util;
+
+namespace Raven.Client.Documents.Session
+{
+    public abstract class ClusterSessionBase : AdvancedSessionExtentionBase
+    {
+        protected ClusterSessionBase(InMemoryDocumentSessionOperations session) : base(session)
+        {
+            if (session.TransactionMode != TransactionMode.ClusterWide)
+            {
+                throw new InvalidOperationException($"This function is part of cluster transaction session, in order to use it you have to open the Session with '{nameof(TransactionMode.ClusterWide)}' option.");
+            }
+        }
+
+        public void CreateCompareExchangeValue<T>(string key, T item)
+        {
+            if (Session.StoreCompareExchange == null)
+                Session.StoreCompareExchange = new Dictionary<string, (object Item, long Index)>();
+
+            EnsureNotDeleted(key);
+
+            Session.StoreCompareExchange.Add(key, (item, 0));
+        }
+
+        public void UpdateCompareExchangeValue<T>(CompareExchangeValue<T> item)
+        {
+            EnsureNotDeleted(item.Key);
+
+            Session.StoreCompareExchange[item.Key] = (item.Value, item.Index);
+        }
+
+        public void DeleteCompareExchangeValue<T>(CompareExchangeValue<T> item)
+        {
+            EnsureNotStored(item.Key);
+
+            if (Session.DeleteCompareExchange == null)
+                Session.DeleteCompareExchange = new Dictionary<string, long>();
+
+            if (Session.DeleteCompareExchange.ContainsKey(item.Key) == false)
+                Session.DeleteCompareExchange.Add(item.Key, item.Index);
+        }
+
+        public void DeleteCompareExchangeValue(string key, long index)
+        {
+            EnsureNotStored(key);
+
+            if (Session.DeleteCompareExchange == null)
+                Session.DeleteCompareExchange = new Dictionary<string, long>();
+            
+            if (Session.DeleteCompareExchange.ContainsKey(key) == false)
+                Session.DeleteCompareExchange.Add(key, index);
+        }
+
+        protected Task<CompareExchangeValue<T>> GetCompareExchangeValueAsyncInternal<T>(string key)
+        {
+            try
+            {
+                return Session.Operations.SendAsync(new GetCompareExchangeValueOperation<T>(key));
```

Comment:
This is wrong. You want to add additional exception handling, but you forgot to `await`

Reply:
OK, I will simply to return the Task